lib/internal/webstreams/readablestream.js

```diff
@@ -0,0 +1,2689 @@
+  }
+
+  /**
+   * @readonly
+   * @type {boolean}
+   */
+  get locked() {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    return isReadableStreamLocked(this);
+  }
+
+  /**
+   * @param {any} [reason]
+   * @returns { Promise<void> }
+   */
+  cancel(reason) {
+    if (!isReadableStream(this))
+      return PromiseReject(new ERR_INVALID_THIS('ReadableStream'));
+    if (isReadableStreamLocked(this)) {
+      return PromiseReject(
+        new ERR_INVALID_STATE.TypeError('ReadableStream is locked'));
+    }
+    return readableStreamCancel(this, reason);
+  }
+
+  /**
+   * @param {{
+   *   mode? : "byob"
+   * }} [options]
+   * @returns {ReadableStreamReader}
+   */
+  getReader(options = {}) {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    validateObject(options, 'options', { nullable: true, allowFunction: true });
+    const mode = options?.mode;
+
+    if (mode === undefined)
+      return new ReadableStreamDefaultReader(this);
+
+    if (`${mode}` !== 'byob')
+      throw new ERR_INVALID_ARG_VALUE('options.mode', mode);
+    return new ReadableStreamBYOBReader(this);
+  }
+
+  /**
+   * @param {ReadableWritablePair} transform
+   * @param {StreamPipeOptions} [options]
+   * @returns {ReadableStream}
+   */
+  pipeThrough(transform, options) {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    const readable = transform?.readable;
+    if (!isReadableStream(readable)) {
+      throw new ERR_INVALID_ARG_TYPE(
+        'transform.readable',
+        'ReadableStream',
+        readable);
+    }
+    const writable = transform?.writable;
+    if (!isWritableStream(writable)) {
+      throw new ERR_INVALID_ARG_TYPE(
+        'transform.writable',
+        'WritableStream',
+        writable);
+    }
+
+    // The web platform tests require that these be handled one at a
+    // time and in a specific order. options can be null or undefined.
+    const preventAbort = options?.preventAbort;
+    const preventCancel = options?.preventCancel;
+    const preventClose = options?.preventClose;
+    const signal = options?.signal;
+
+    if (signal !== undefined && signal?.[kAborted] === undefined)
+      throw new ERR_INVALID_ARG_TYPE('options.signal', 'AbortSignal', signal);
+
+    if (isReadableStreamLocked(this))
+      throw new ERR_INVALID_STATE.TypeError('The ReadableStream is locked');
+    if (isWritableStreamLocked(writable))
+      throw new ERR_INVALID_STATE.TypeError('The WritableStream is locked');
+
+    const promise = readableStreamPipeTo(
+      this,
+      writable,
+      !!preventClose,
+      !!preventAbort,
+      !!preventCancel,
+      signal);
+    setPromiseHandled(promise);
+
+    return readable;
+  }
+
+  /**
+   * @param {WritableStream} destination
+   * @param {StreamPipeOptions} [options]
+   * @returns {Promise<void>}
+   */
+  pipeTo(destination, options = {}) {
```

Comment:
this function sometimes throws synchronously, and sometimes returns a rejected promise.

i think all the throws need to become "return Promise.reject()"

Reply:
The synchronous throws are all wrapped in a try catch that returns a rejected promise.